statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/functions/FlinkStateDelayedMessagesBuffer.java

```diff
@@ -26,41 +33,121 @@
 final class FlinkStateDelayedMessagesBuffer implements DelayedMessagesBuffer {
 
   static final String BUFFER_STATE_NAME = "delayed-messages-buffer";
+  static final String INDEX_STATE_NAME = "delayed-message-index";
 
   private final InternalListState<String, Long, Message> bufferState;
+  private final MapState<String, Long> messageIdToTimestamp;
 
   @Inject
   FlinkStateDelayedMessagesBuffer(
-      @Label("delayed-messages-buffer-state")
-          InternalListState<String, Long, Message> bufferState) {
+      @Label("delayed-messages-buffer-state") InternalListState<String, Long, Message> bufferState,
+      @Label("delayed-message-index") MapState<String, Long> messageIdToTimestamp) {
     this.bufferState = Objects.requireNonNull(bufferState);
+    this.messageIdToTimestamp = Objects.requireNonNull(messageIdToTimestamp);
   }
 
   @Override
-  public void add(Message message, long untilTimestamp) {
-    bufferState.setCurrentNamespace(untilTimestamp);
+  public void forEachMessageAt(long timestamp, Consumer<Message> fn) {
     try {
-      bufferState.add(message);
+      forEachMessageThrows(timestamp, fn);
     } catch (Exception e) {
-      throw new RuntimeException("Error adding delayed message to state buffer: " + message, e);
+      throw new IllegalStateException(e);
     }
   }
 
   @Override
-  public Iterable<Message> getForTimestamp(long timestamp) {
-    bufferState.setCurrentNamespace(timestamp);
-
+  public OptionalLong removeMessageByCancellationToken(String token) {
     try {
-      return bufferState.get();
+      return remove(token);
     } catch (Exception e) {
-      throw new RuntimeException(
-          "Error accessing delayed message in state buffer for timestamp: " + timestamp, e);
+      throw new IllegalStateException("Failed clearing a message with id " + token, e);
```

Comment:
Maybe change `id` in the error message to `cancellation token`, to be consistent and not confuse users.